In [4]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [5]:
df = pd.read_csv("in-vehicle-coupon-recommendation.csv")

In [6]:
categorical_features = df.select_dtypes(include=['object']).columns
df_encoded = df.copy()

In [ ]:
# Applying Label Encoding to ordinal categorical features
label_encoders = {}
ordinal_features = ['expiration', 'maritalStatus']
for col in ordinal_features:
    le = LabelEncoder()
    df_encoded[col] = le.fit_transform(df[col])
    label_encoders[col] = le


# Apply One-Hot Encoding to nominal categorical features

In [8]:
df_encoded = pd.get_dummies(df_encoded, columns=[col for col in categorical_features if col not in ordinal_features], drop_first=True)


Model performance analysis

In [9]:
X = df.drop(columns=['Y']) 
y = df['Y'] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

ValueError: could not convert string to float: 'No Urgent Place'

In [11]:
X = df_encoded.drop(columns=['Y']) 
y = df_encoded['Y'] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

accuracy_encoded = accuracy_score(y_test, y_pred)
print(accuracy_encoded)

0.7445802128498227


# Observations
1. Encoding categorical variables increased feature dimensionality.
   - One-Hot Encoding expanded the dataset but avoids ordinal misinterpretation.
   - Label Encoding preserved ordinal relationships.
2. Model accuracy after encoding: 0.7445802128498227